### PR TITLE
Refactor the account info output type to be compatible with `AIP-62` standard

### PR DESCRIPTION
**Pull request by 0xmaayan**
To support the different Public Key types and for developers to easily be able to pass the account info down to SDK functions, changing the Account Info return type to be compatible with AIP-62

An updated version with this change here: https://aptos-labs.github.io/aptos-wallet-adapter/nextjs-example-testing/

Will be released with a major version update

OLD
```
export type AccountInfo = {
  address: string;
  publicKey: string | string[];
  minKeysRequired?: number;
  ansName?: string | null;
};
```

NEW
```
export type AccountInfoOutput = {
  address: AccountAddress;
  publicKey: PublicKey | PublicKey[]; // PublicKey[] for backward compatibility
  minKeysRequired?: number;
  ansName?: string | null;
};
```

This way, one can do
```
  const signWithSDK = async () => {
    if (!account) return;

    // make sure it is not an array of public keys
    if (Array.isArray(account.publicKey)) return;

    const transaction = await aptosClient(network).transaction.build.simple({
      sender: account.address,
      data: {
        function: "0x1::coin::transfer",
        typeArguments: [APTOS_COIN],
        functionArguments: [account.address.toString(), 1], // 1 is in Octas
      },
    });
    const response = await aptosClient(network).transaction.simulate.simple({
      signerPublicKey: account.publicKey,
      transaction,
    });
  };
```

### CODE REVIEW

File: packages/wallet-adapter-core/src/WalletCore.ts [hunk 1]
```diff
@@ -783,6 +759,7 @@ export class WalletCore extends EventEmitter<WalletCoreEvents> {
       } else {
         account = await this.walletCoreV1.connect(selectedWallet);
       }
+      console.log("connectWallet account", account);
```
Comment (GhostWalker562) [inline packages/wallet-adapter-core/src/WalletCore.ts]: Probably remove?

File: packages/wallet-adapter-core/src/WalletCore.ts [hunk 2]
```diff
@@ -687,9 +642,30 @@ export class WalletCore extends EventEmitter<WalletCoreEvents> {
...
+      if (!this._account) return null;
+      return {
+        address:
+          typeof this._account.address === "string"
+            ? AccountAddress.from(this._account.address)
+            : this._account.address,
+        publicKey:
+          // for backward compatibility, if the account public key is of type `string` convert it to Ed25519PublicKey
+          typeof this._account.publicKey === "string"
+            ? new Ed25519PublicKey(this._account.publicKey)
```
Comment (GhostWalker562) [inline packages/wallet-adapter-core/src/WalletCore.ts:656]: Is it possible to adjust this a bit so that we can account for SingleKey's as well?

If an outdated wallet shares any other public key that is not Ed25519, this will probably break

File: packages/wallet-adapter-core/src/LegacyWalletPlugins/types.ts [hunk 3]
```diff
@@ -74,7 +79,7 @@ export declare interface WalletCoreEvents {
   readyStateChange(wallet: Wallet): void;
   standardWalletsAdded(wallets: Wallet | AptosStandardSupportedWallet): void;
   networkChange(network: NetworkInfo | null): void;
-  accountChange(account: AccountInfo | null): void;
+  accountChange(account: AccountInfo | StandardAccountInfoInput | null): void;
```
Comment (GhostWalker562) [inline packages/wallet-adapter-core/src/LegacyWalletPlugins/types.ts:82]: Do you think it might be better to just share the `AccountInfoOutput` instead?

Comment (0xmaayan) [reply packages/wallet-adapter-core/src/LegacyWalletPlugins/types.ts:82]: This is tricky as the adapter still supports legacy and new standard types, so the account the adapter gets back from the wallet can be of both types. Changing it gets us into a type-rabbit-hole, so we simply change the actual input type the adapter returns.

In the near future, we should probably support _only_ AIP-62 wallets

File: packages/wallet-adapter-core/src/WalletCore.ts (hunk 2 above)
Comment (0xmaayan) [reply packages/wallet-adapter-core/src/WalletCore.ts:656]: I'd argue, if you are an outdated wallet - you should update.

Most of the used / updated wallets are already on AIP-62 so they should never really get here, we keep it for the small amount of wallets the adapter still needs to support - which we should probably deprecate in the near future.

In the adapter context, if the `public key` is of type string, we assume it is an old standard which most chances is just Ed255519.

**Review by GhostWalker562 — APPROVED** [review]

File: packages/wallet-adapter-core/src/WalletCore.ts (hunk 2 above)
Comment (hardsetting) [reply packages/wallet-adapter-core/src/WalletCore.ts:656]: Agreed, the old interface does not support non-ed25519 public keys, so if they were serializing other types of keys they were probably doing something wrong

File: packages/wallet-adapter-core/src/LegacyWalletPlugins/types.ts (hunk 3 above)
Comment (hardsetting) [reply packages/wallet-adapter-core/src/LegacyWalletPlugins/types.ts:82]: I think we should use `AccountInfo` from `@aptos-labs/wallet-standard` here, that uses the `PublicKey` type.

If I understand correctly, this is an event that can be listened to externally by the dapp.
Since we're bumping the major, we might as well clean up the interface and have the "right" type here.

In case of legacy plugins, we should be able to adapt the return types to `AccountInfo` so that the dapp sees one consistent return type.

EDIT: [here's](https://github.com/aptos-labs/aptos-wallet-adapter/blob/bde811255e523d050844885312cc2f2e0d73c91d/packages/wallet-adapter-core/src/WalletCore.ts#L1139) a good spot where we could normalize

File: packages/wallet-adapter-core/src/WalletCore.ts [hunk 4]
```diff
@@ -131,7 +131,7 @@ export class WalletCore extends EventEmitter<WalletCoreEvents> {
   private _wallet: Wallet | null = null;
 
   // Current connected account
-  private _account: AccountInfo | null = null;
+  private _account: AccountInfo | StandardAccountInfoInput | null = null;
```
Comment (hardsetting) [inline packages/wallet-adapter-core/src/WalletCore.ts:134]: Same here, let's just use `AccountInfo` from the standard. We can convert it to the right type before setting its value

File: packages/wallet-adapter-core/src/WalletCore.ts [hunk 5]
```diff
@@ -465,7 +465,7 @@ export class WalletCore extends EventEmitter<WalletCoreEvents> {
    * @param account An account
    */
   private ensureAccountExists(
-    account: AccountInfo | null
+    account: AccountInfo | StandardAccountInfoInput | null
```
Comment (hardsetting) [inline packages/wallet-adapter-core/src/WalletCore.ts:468]: same here

File: packages/wallet-adapter-core/src/WalletCore.ts [hunk 6]
```diff
@@ -541,57 +541,12 @@ export class WalletCore extends EventEmitter<WalletCoreEvents> {
...
-        return;
-      }
-    }
-
-    // account is of type `AccountInfo`
-    this._account = { ...(account as AccountInfo) };
-    return;
+  setAccount(account: AccountInfo | StandardAccountInfo | null): void {
+    this._account = account;
   }
```
Comment (hardsetting) [inline packages/wallet-adapter-core/src/WalletCore.ts:550]: This would be a good spot where we support both `AccountInfo`s (the old internal type, and the standard type) as inputs, and normalize it to the standard one.

File: packages/wallet-adapter-core/src/WalletCore.ts [hunk 7]
```diff
@@ -687,9 +642,30 @@ export class WalletCore extends EventEmitter<WalletCoreEvents> {
    * @return account info
    * @throws WalletAccountError
    */
-  get account(): AccountInfo | null {
+  get account(): AccountInfoOutput | null {
```
Comment (hardsetting) [inline packages/wallet-adapter-core/src/WalletCore.ts:645]: Let's just return `AccountInfo` from the wallet standard package

File: packages/wallet-adapter-core/src/WalletCore.ts [hunk 8]
```diff
@@ -687,9 +642,30 @@ export class WalletCore extends EventEmitter<WalletCoreEvents> {
...
+        publicKey:
+          // for backward compatibility, if the account public key is of type `string` convert it to Ed25519PublicKey
+          typeof this._account.publicKey === "string"
+            ? new Ed25519PublicKey(this._account.publicKey)
+            : // for backward compatibility, if the account public key is of type `string[]` convert each string to Ed25519PublicKey
+              Array.isArray(this._account.publicKey) &&
+                this._account.publicKey.every(
+                  (item) => typeof item === "string"
+                )
+              ? this._account.publicKey.map((key) => new Ed25519PublicKey(key))
```
Comment (hardsetting) [inline packages/wallet-adapter-core/src/WalletCore.ts:662]: Ideally we should convert `string[]` and `minKeysRequired` to `MultiEd25519PublicKey` that extends `PublicKey`

File: apps/nextjs-example/src/app/page.tsx [hunk 9]
```diff
@@ -147,7 +150,7 @@ function WalletSelection() {
 }
 
 interface WalletConnectionProps {
-  account: AccountInfo | null;
+  account: AccountInfoOutput | null;
```
Comment (hardsetting) [inline apps/nextjs-example/src/app/page.tsx:153]: let's use `AccountInfo` from the standard

File: packages/wallet-adapter-core/src/utils/walletSelector.ts [hunk 10]
```diff
@@ -44,9 +45,9 @@ export function isInstallRequired(wallet: AnyAptosWallet) {
 }
 
 /** Truncates the provided wallet address at the middle with an ellipsis. */
-export function truncateAddress(address: string | undefined) {
+export function truncateAddress(address: AccountAddress | undefined) {
```
Comment (hardsetting) [inline packages/wallet-adapter-core/src/utils/walletSelector.ts:48]: small nit: we could use `AccountAddressInput`, but nbd

**Review by hardsetting — CHANGES_REQUESTED** [review]
Thanks for making this change!
As per my comments, I think we should try using `AccountInfo` from the standard :)

**Comment by hardsetting** [comment]
By quickly looking at the code, I think we could do this:
- define a `normalizeAccountInfo` that normalizes any `AccountInfo` to the standard's one. Here you would put the code that converts `string` to `Ed25519PublicKey` and `(string[], minKeysRequired)` to `MultiEd25519PublicKey`
- when calling `connectWallet` [here](https://github.com/aptos-labs/aptos-wallet-adapter/blob/bde811255e523d050844885312cc2f2e0d73c91d/packages/wallet-adapter-core/src/WalletCore.ts#L781-L786), we normalize the account info before calling `setAccount`
- when we receive an "accountChange" event [here](https://github.com/aptos-labs/aptos-wallet-adapter/blob/bde811255e523d050844885312cc2f2e0d73c91d/packages/wallet-adapter-core/src/WalletCore.ts#L1139), we normalize the account info before `setAccount` and `emit`
- `setAccount` should only take the standard's `AccountInfo` as input

basically.. let's normalize as soon as we have a chance to make our life easier :)

**Comment by 0xmaayan** [comment]
closed in favor of https://github.com/aptos-labs/aptos-wallet-adapter/pull/464
